### PR TITLE
Ulimits are now merged into extended services

### LINKED ISSUE #2607: ulimits not merged when used with extends
Hey there,

I try to tweak the `max locked memory` within a container, which could be done by starting it with this options:
```
--ulimit memlock="68719476736"
```
This results in `max locked memory       (kbytes, -l) 67108864`.

If I am not mistaken ([code](https://github.com/docker/compose/blob/a2d2915a64c887ddc4bfa2a2e41bfb65eb4f86ad/compose/service.py#L936-L946)) the compose script works with arbitrary key/value pairs.

If I start a container like this:
```
hpcg2:
    extends:
      file: base.yml
      service: hpcg
    hostname: hpcg2
    container_name: hpcg2
    environment:
     - constraint:node==venus002
    ulimits:
      nofile:
        soft: 20000
        hard: 40000
      memlock:
        soft: 68719476736
        hard: 68719476736
```
The `.HostConfig.Ulimits` stays empty.
```
[root@venus001 hpcg]# docker inspect -f '{{ .HostConfig.Ulimits }}' hpcg2
[]
```

Started like this:
```
docker run -d --net=global \
                -v "/scratch/:/scratch/" \
                --dns=127.0.0.1 \
                --privileged \
                -e DC_NAME=dc1 -e CONSUL_CLUSTER_IPS=consul1,consul2 -e COLLECT_METRICS=true \
                -e SENSU_CHECK_IB=true \
                --hostname=hpcg${1} \
                --name=hpcg${1} \
                -e constraint:node==venus00${1} \
                --ulimit memlock="68719476736" \
        qnib/hpcg tail -f /dev/null
```
It's fine:
```
[root@venus001 hpcg]# docker inspect -f '{{ .HostConfig.Ulimits }}' hpcg3
[memlock=68719476736:68719476736]
```

I use 1.5.1:
```
[root@venus001 hpcg]# docker-compose version
docker-compose version: 1.5.1
docker-py version: 1.5.0
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1e 11 Feb 2013
[root@venus001 hpcg]#
```
and 1.5.2:
```
[root@venus001 hpcg]# compose version
docker-compose version 1.5.2, build 7240ff3
docker-py version: 1.5.0
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1e 11 Feb 2013
[root@venus001 hpcg]#
```

Any hint why the ulimits do not show up?
EDIT: Should docker-py be in lockstep with docker-compose? Since ulimit came with 1.5.1...

### PR DESCRIPTION
**Pull request by dbonev**
Fixes #2607

The issue was that the ```ulimits``` were not merged in extended services.

Signed-off-by: Dimitar Bonev <dimitar.bonev@gmail.com>

### CODE REVIEW

File: tests/unit/config/config_test.py [hunk 1]
```diff
@@ -346,6 +346,11 @@ def test_load_with_multiple_files_and_extends_in_override_file(self):
         tmpdir.join('common.yml').write("""
             base:
               labels: ['label=one']
+              ulimits:
+                nproc: 65535
+                memlock:
+                    soft: 1024
+                    hard: 2048
```
Comment (dnephin) [inline tests/unit/config/config_test.py]: I think this should be added as a separate case in the `ExtendsTest` TestCase. This test is unrelated

File: compose/config/config.py [hunk 2]
```diff
@@ -725,6 +726,20 @@ def split_label(label):
         return label, ''
 
 
+def merge_ulimits(base, override):
+    ulimits = parse_ulimits(base)
+    ulimits.update(parse_ulimits(override))
+    return ulimits
```
Comment (dnephin) [inline compose/config/config.py]: I swear that we'd already fixed this a couple times, but apparently not, and now I can't find any open PRs with the fix.

I think all of these merge functions (`merge_labels`, `merge_environment`) should be just a single `merge_mapping` function that takes a function to parse the base and override.

Would you mind making that change instead of adding a new one?

**Comment by dnephin** [comment]
Thanks for diagnosing the issue and fixing it!  A couple small requests, but mostly looks good.

**Comment by dbonev** [comment]
@dnephin Thanks for the comments.
- refactored the ```merge_``` methods into ```merge_mapping``` accepting a ```parse``` function
- created a separated test for this behavior

Please let me know if something doesn't look right.

**Comment by dnephin** [comment]
LGTM!

**Comment by aanand** [comment]
LGTM
